File: frontend/src/api/admin/customer-codes/bulk-delete.ts
/**
 * `POST /admin/customer-codes/bulk-delete` — delete multiple customer codes by id.
 */

import { postData } from "@/api/client"

export interface BulkDeleteCustomerCodesRequest {
  ids: string[]
}

export interface BulkDeleteCustomerCodesResponse {
  deleted: number
}

/**
 * Delete the customer codes with the provided ids.
 *
 * @param payload - Object containing the list of ids to delete.
 * @returns The number of documents actually deleted.
 */
export async function bulkDeleteCustomerCodes(
  payload: BulkDeleteCustomerCodesRequest,
): Promise<number> {
  const result = await postData<BulkDeleteCustomerCodesResponse>(
    "/admin/customer-codes/bulk-delete",
    payload,
  )
  return result.deleted
}

File: backend/app/services/customer_code/bulk_delete.py
"""Customer code service: bulk-delete selected records by id.

Deletes up to 100 documents in one MongoDB operation. Invalid/unknown ids are
ignored; the caller receives the number of documents actually removed. A single
audit event is emitted for the batch.
"""

from __future__ import annotations

from beanie import PydanticObjectId

from ...models.customer_code import CustomerCode
from ..audit.events import audit_customer_code_event


async def bulk_delete_customer_codes(
    ids: list[str],
    *,
    actor_email: str | None,
) -> int:
    """Delete all valid ``CustomerCode`` documents whose id is in ``ids``.

    Args:
        ids: ObjectId hex strings. Duplicates and invalid hex strings are
             ignored silently.
        actor_email: Admin email for the audit event.

    Returns:
        Number of documents actually deleted.
    """
    valid_ids = []
    for raw in ids:
        try:
            valid_ids.append(PydanticObjectId(raw))
        except Exception:
            continue

    if not valid_ids:
        return 0

    result = await CustomerCode.find({"_id": {"$in": valid_ids}}).delete()
    deleted_count = getattr(result, "deleted_count", 0) or 0

    await audit_customer_code_event(
        "customer_code.bulk_deleted",
        f"Deleted {deleted_count} customer code(s)",
        actor_email=actor_email,
        extra={"count": deleted_count, "requested": len(ids), "ids": ids},
    )

    return deleted_count
